### DISCUSSION TITLE
What cookie(s) should a client request include to authenticate it?

### DISCUSSION THREAD

**Discussion by dsaghliani** [maxcountryman/axum-login#45] [Q&A] [answered]
Sorry about the newbie question. It's my first time writing auth.

When a user is logged in on the server, it should respond with the data necessary for authentication, right? Data that should be stored on the client as a cookie so that it's attached with every subsequent request.

Since `axum-login` uses `tower_http`'s `RequireAuthorizationLayer`, I assume it checks the [`Authorization` header](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Authorization), as it says in the docs. But what value should I assign that?

Looking at `axum-login`'s source code, I see that it attempts to fetch the `Option<User>` extension in the request, where `User: AuthUser<Role>`.

```rs
impl<User, ReqBody, ResBody, Role> AuthorizeRequest<ReqBody> for Login<User, ResBody, Role>
where
    Role: PartialOrd + PartialEq + Clone + Send + Sync + 'static,
    User: AuthUser<Role>,
    ResBody: HttpBody + Default,
{
    type ResponseBody = ResBody;

    fn authorize(
        &mut self,
        request: &mut Request<ReqBody>,
    ) -> Result<(), Response<Self::ResponseBody>> {
        let user = request
            .extensions()
            .get::<Option<User>>()
            .expect("Auth extension missing. Is the auth layer installed?");

        match user {
            Some(user) if self.role_bounds.contains(user.get_role()) => {
                let user = user.clone();
                request.extensions_mut().insert(user);

                Ok(())
            }

            _ => {
                let unauthorized_response = Response::builder()
                    .status(http::StatusCode::UNAUTHORIZED)
                    .body(Default::default())
                    .unwrap();

                Err(unauthorized_response)
            }
        }
    }
}
```

Is that what I'm supposed to include in the `Authorization` header? Does that mean I'm expected to pass around a struct with fields `id`, `role`, and `password_hash` with every request? Is that safe?

**Comment by maxcountryman** [comment, marked as answer]
axum-login uses axum-sessions which in turn sets its own cookie (by default this is axum.sid but can be configured to be called whatever you like).

While the docs for RequireAuthorizationLayer do mention the Authorization header in our implementation that part is not relevant: we instead use the middleware directly to obtain a User if one is present. This works because axum-sessions manages the cookie for us and axum-login provides the user context via that.

So the only thing that needs to be passed back to the server is that cookie, axum.sid. This also means the cookie is sensitive and should never be shared. For that reason the cookie uses a number of default settings that help ensure it cannot be easily stolen.

**Reply by dsaghliani** [reply]
> So the only thing that needs to be passed back to the server is that cookie, axum.sid. This also means the cookie is sensitive and should never be shared. For that reason the cookie uses a number of default settings that help ensure it cannot be easily stolen.

That's cool. Cookies are passed automatically, right? That's weird, because it doesn't seem to work for me. The server keeps returning `401: Unauthorized`, even though the `axum.sid` cookie is present.

How would you recommend debugging this? Is there a way to see into the memory store at runtime? Or into whatever code authorizes requests?

**Reply by maxcountryman** [reply]
A guess in the dark: is secure set to true with axum-sessions? By default it will be and locally browsers like Safari will silently drop the cookie without proper SSL/TLS. You can try setting that to false, but for production I would strongly encourage setting it to true.

As far as debugging, I would like to instrument more with the tracing crate so we have better visibility into the inner workings of the crate. I haven't had time to look more into that tho.

**Reply by dsaghliani** [reply]
> A guess in the dark: is secure set to true with axum-sessions? By default it will be and locally browsers like Safari will silently drop the cookie without proper SSL/TLS. You can try setting that to false, but for production I would strongly encourage setting it to true.

Good clue! Unfortunately, that doesn't seem to be it. I set it to `false` and ensured the cookie is included in the request header (at least, it is in the dev tools), and I know the verification works because the server returns a `Redirect` response, which only happens in case of success. But then it's still `401: Unauthorized`.

**Reply by dsaghliani** [reply]
So, I did a little digging and edited the `Service` implementation of `AuthService`, printing the value of `auth_cx.get_user()` in `call()`:

```rs
fn call(&mut self, mut request: Request<ReqBody>) -> Self::Future {
    let state = self.state.clone();
    let inner = self.inner.clone();

    let mut inner = std::mem::replace(&mut self.inner, inner);
    Box::pin(async move {
        let Extension(session_handle): Extension<SessionHandle> = request
            .extract_parts()
            .await
            .expect("Session extension missing. Is the session layer installed?");

        let mut auth_cx = AuthContext::new(session_handle, state.store, state.key);
        match auth_cx.get_user().await {
            Ok(user) => {
                auth_cx.current_user = user;

+               if let Some(user) = &auth_cx.current_user {
+                   println!(
+                       "Inserted user with the ID: {:#?}\nand the password hash: {:#?}",
+                       user.get_id(),
+                       user.get_password_hash().expose_secret()
+                   );
+               } else {
+                   println!("`user` is `None`.");
+               }

                request.extensions_mut().insert(auth_cx.clone());
                request.extensions_mut().insert(auth_cx.current_user);

                inner.call(request).await
            }

            Err(err) => {
                tracing::error!("Could not get user: {:?}", err);
                let response = Response::builder()
                    .status(http::StatusCode::INTERNAL_SERVER_ERROR)
                    .body(Default::default())
                    .unwrap();

                Ok(response)
            }
        }
    })
}
```

In my own project, I added a line that prints `auth_ctx.current_user` after `auth_ctx.login()` is called. As expected, that returns `Some(user)`. However, the `println!()` in `call()` still prints that `user` is `None`:

```
2022-12-31T05:44:39.746399Z DEBUG hyper::proto::h1::conn: incoming body completed
`user` is `None`.
2022-12-31T05:44:39.752127Z  INFO sqlx::query: SELECT typname, typtype, typcategory, …; rows affected: 0, rows returned: 1, elapsed: 1.046ms

SELECT
  typname,
  typtype,
  typcategory,
  typrelid,
  typelem,
  typbasetype
FROM
  pg_catalog.pg_type
WHERE
  oid = $1
  
2022-12-31T05:44:39.754574Z  INFO sqlx::query: SELECT enumlabel FROM pg_catalog.pg_enum …; rows affected: 3, rows returned: 3, elapsed: 1.161ms

SELECT
  enumlabel
FROM
  pg_catalog.pg_enum
WHERE
  enumtypid = $1
ORDER BY
  enumsortorder
  
2022-12-31T05:44:39.765216Z  INFO sqlx::query: SELECT id, password_hash, role …; rows affected: 0, rows returned: 1, elapsed: 16.699ms

SELECT
  id,
  password_hash,
  role AS "role: _"
FROM
  "user"
WHERE
  email = $1
  
2022-12-31T05:44:40.137914Z DEBUG prompt_exchange::accounts::login: Current user: Some(
    AuthUser {
        id: 8,
        password_hash: "$argon2id$v=19$m=4096,t=3,p=1$lVT4JqLbJ5OxSdo9ANTLeA$ykwYvtaljBTqHjj4eN9kS8jC7XosLD5RBjWdSK4sS6c",
        role: User,
    },
)
2022-12-31T05:44:40.138277Z DEBUG hyper::proto::h1::io: flushed 127 bytes
2022-12-31T05:44:40.427295Z DEBUG hyper::proto::h1::io: parsed 13 headers
2022-12-31T05:44:40.427444Z DEBUG hyper::proto::h1::conn: incoming body is empty
`user` is `None`.
2022-12-31T05:44:40.428097Z DEBUG hyper::proto::h1::io: flushed 82 bytes
2022-12-31T05:44:40.608915Z DEBUG hyper::proto::h1::io: parsed 13 headers
2022-12-31T05:44:40.608961Z DEBUG hyper::proto::h1::conn: incoming body is empty
`user` is `None`.
2022-12-31T05:44:40.609507Z DEBUG hyper::proto::h1::io: flushed 82 bytes
^[2022-12-31T05:46:36.008002Z DEBUG hyper::proto::h1::conn: read eof
```

What could be the reason for this?

**Reply by maxcountryman** [reply]
It's hard to help without a runnable example that demonstrates the problem. Can you put together a minimal example in a public repo that we can look at?

**Reply by dsaghliani** [reply]
You're right. Here: https://github.com/dsaghliani/axum-login-minimal-example

**Reply by maxcountryman** [reply]
Unfortunately that has a path dependency, maud, which means I can't run it locally.

I'd strongly suggest removing as many deps as possible and minimizing the surface area as much as possible. It will make it much easier to pin point any potential issues. I'd probably start with one of the examples in this repo and see if you can reproduce your issue with those.

**Reply by dsaghliani** [reply]
I'll do that. In the meantime, if you're willing to take a look, I've removed maud and replaced it with plain HTML.

**Reply by dsaghliani** [reply]
I figured out the problem: I was using a dummy `MemoryStore` and had not populated it with users. It appears I misunderstood how `axum-login` works. I thought, when I called `AuthContext::login()`, which sets `self.current_user = Some(user.clone())`, that `current_user` would be persisted between handlers, but I guess it's just for use in the same handler (that called `login()`)? And the middleware queries the database and refetches `User: AuthUser` with every request?

**Reply by maxcountryman** [reply]
Glad you figured it out.

The second part of what you said is true: the user needs to be queryable and if it's not in the store then you won't be able to access it.

This supports consistency when a user is changed or deleted. Otherwise your application might be accessing a stale copy which could be dangerous.

You can see how this behavior is implemented by checking out the login middleware itself which attaches the User to the request.

**Reply by dsaghliani** [reply]
That explains it. While we're on the topic of auth, do you suggest using Postgres or Redis as the session store, provided auth data is the only thing stored in sessions? Since the auth service has to read from the database anyway, could it be somehow be optimized to grab the session data along the way to avoid 2 reads? Or is that a silly question?

**Reply by maxcountryman** [reply]
Great question. I'm using Postgres to minimize the complexity of my infra. Redis is the more traditional answer but I'm not convinced it's actually better for most applications.

As far as the query optimization question, that would be nice but I don't think there's any practical way to do that. Even if we assume the session and the user are in the same store, we don't have a way of making axum-session aware of axum-login or vice versa. (We could probably find some way of doing this but I suspect it would be awkward--as it stands each is its own module and the way I can think of to enable that would be to make e.g. axum-login also responsible for querying the session.)
